server/auth_function/test/util.py

```diff
@@ -0,0 +1,135 @@
+    )
+
+    cursor.execute(
+        raw_query, (lambda_function.USER_TYPE_CODE_DICT[idp_name], user_name)
+    )
+
+
+def initial_user(db_pg_transaction, cognito_event):
+    cursor = db_pg_transaction.cursor()
+
+    raw_query = """INSERT INTO app_fam.fam_user
+        (user_type_code, user_name, user_guid,
+        create_user, create_date, update_user, update_date)
+        VALUES( %s, %s, %s,
+        CURRENT_USER, CURRENT_DATE, CURRENT_USER, CURRENT_DATE);"""
+    # print(f"query is\n:{raw_query}")
+
+    idp_name = cognito_event["request"]["userAttributes"].get("custom:idp_name")
+    # set "username" to be "custom:idp_user_id" when "custom:idp_username" not exists for bcsc user
+    user_name = (
+        cognito_event["request"]["userAttributes"].get("custom:idp_username")
+        if cognito_event["request"]["userAttributes"].get("custom:idp_username")
+        else cognito_event["request"]["userAttributes"].get("custom:idp_user_id")
+    )
+    # For Insert statement, pass parameters as .execute()'s second arguments so they get proper sanitization.
+    cursor.execute(
+        raw_query,
+        (
+            lambda_function.USER_TYPE_CODE_DICT[idp_name],
+            user_name,
+            cognito_event["request"]["userAttributes"].get("custom:idp_user_id"),
+        ),
+    )
+
+
+def create_test_fam_cognito_client(db_pg_transaction, cognito_event):
+    client_id = cognito_event["callerContext"]["clientId"]
+    cursor = db_pg_transaction.cursor()
+    # Set up expected client in DB
+    # cursor = db_transaction.cursor()
+    raw_query = """
+    insert into app_fam.fam_application_client
+        (cognito_client_id,
+        application_id,
+        create_user,
+        update_user)
+    values(
+        %s,
+        (select application_id from app_fam.fam_application
+            where application_name = 'FAM'),
+        CURRENT_USER,
+        CURRENT_USER)
+    """
+    # For Insert statement, pass parameters as .execute()'s second arguments so they get proper sanitization.
+    cursor.execute(raw_query, [client_id])
+
+
+def create_user_role_xref_record(db_pg_transaction, test_user_properties):
+    cursor = db_pg_transaction.cursor()
+    raw_query = """
+    insert into app_fam.fam_user_role_xref
+        (user_id,
+        role_id,
+        create_user,
+        update_user)
+    VALUES (
+        (select user_id from app_fam.fam_user where
+            user_name = %s
+            and user_type_code = %s),
+        (select role_id from app_fam.fam_role where
+            role_name = %s),
+        CURRENT_USER,
+        CURRENT_USER
+    )
+    """
+    # For Insert statement, pass parameters as .execute()'s second arguments so they get proper sanitization.
+    cursor.execute(
+        raw_query,
+        (
+            test_user_properties.get("idp_username"),
+            test_user_properties.get("idp_type_code"),
+            TEST_ROLE_NAME,
+        ),
+    )
```

Comment:
Just curious, were these functions previous from `conftest.py` not easy to use (despite you refactored a little on code)? 
All of these functions migrated into utils.py seems very much like dependencies before running tests, which is one thing pytest's "**@pytest.fixture**" is good at and convenient to use. And fixture can contain other fixtures, the function in test does not need to call all of them(the setup/dependencies).

Reply:
oh yeah, initially I need to pass some parameter into the cognito_event fixture, but you're right, let me try again, maybe it's now working.

Reply:
Hi Ian, you're right, even we pass parameters the fixture contains another fixture still works as expected, I switched back, thanks so much for noticing this!!